### PR TITLE
[MLIR] Remove validations that have been moved into MLIR.

### PR DESCRIPTION
**Pull request by krzysz00**
Some of the checks that the MLIR solvers perform can be moved into
MLIR, allowing MLIR to enable support for more types of convolution
without needing to coordinate as many changes with MIOpen. This also
reduces redundancy in the solver applicability checks.

### CODE REVIEW

**Comment by krzysz00** [comment]
This PR depends on [MLIR#298](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/pull/298), so this is here for review and coordination.

**Comment by atamazov** [comment]
@krzysz00 Is it ready for merging? If not, please convert to draft.

File: src/solver/mlir_common.cpp [hunk 1]
```diff
@@ -41,6 +43,30 @@ std::string InsertGToLayout(const std::string& layout, char dim)
     return layout_with_g.insert(index, 1, 'G');
 }
 
+const char* DTypeName(miopenDataType_t ty) {
```
Comment (jerryyin) [inline src/solver/mlir_common.cpp:46]: I know in all scenarios you are returning a string literal with static storage duration. However I'd prefer this to be returning a string so client of this function never worry about the implementation details of this one.

File: src/solver/mlir_common.cpp [hunk 2]
```diff
@@ -41,6 +43,30 @@ std::string InsertGToLayout(const std::string& layout, char dim)
...
+    case miopenDouble:
+        return "fp64";
+    case miopenBFloat16:
+        return "bf16";
+    case miopenInt32:
+        return "i32";
+    case miopenInt8:
+        return "i8";
+    case miopenInt8x4:
+        return "i8x4";
```
Comment (jerryyin) [inline src/solver/mlir_common.cpp:61]: @whchung FYI, we are creating this 1:1 mapping string for data type representations. Right now, all except fp16, fp32 should be rejected by MLIR. To enable a new data type wouldn't require any change to MIOpen in the future.

File: src/solver/mlir_common.cpp [hunk 3]
```diff
@@ -56,10 +82,13 @@ std::string ConstructBuildOptions(const ConvolutionContext& ctx,
     std::string out_layout = InsertGToLayout(CI::GetOutputLayout(ctx), 'C');
 
     std::string mlir_handle;
-    if (is_xdlops)
-        mlir_handle += std::string(" --x2 ") + "1";
+    if (!ctx.Is2d()) {
+        // Future: Remove this once MLIr supports 3D convolutions
+        return mlir_handle;
+    }
```
Comment (jerryyin) [inline src/solver/mlir_common.cpp:88]: Can we move this to MLIR?

**Review by jerryyin — CHANGES_REQUESTED** [review]
Some minor nits. Overall good job. One missing thing is that you'd need to bump the MLIR commit in MIOpen's Dockerfile once your MLIR PR is merged.

File: src/solver/mlir_common.cpp (hunk 3 above)
Comment (krzysz00) [reply src/solver/mlir_common.cpp:88]: Well, if we wanted to add a series of `--*_depth` options to MLIR and then reject on depth != 0, then we could. Is that a preferable refactor.

File: src/solver/mlir_common.cpp (hunk 1 above)
Comment (krzysz00) [reply src/solver/mlir_common.cpp:46]: On the other hand, if we're ever using anything but string literals, we can refactor, and this has the advantage of making it clear that it's meant to be a static string literal.

File: src/solver/mlir_common.cpp (hunk 3 above)
Comment (jerryyin) [reply src/solver/mlir_common.cpp:88]: Can we do an additional check on layout to filter off those that whose length are not 5?

File: src/solver/mlir_common.cpp (hunk 1 above)
Comment (jerryyin) [reply src/solver/mlir_common.cpp:46]: Fine by me.

File: src/solver/mlir_common.cpp (hunk 3 above)
Comment (jerryyin) [reply src/solver/mlir_common.cpp:88]: On a second look you already have it in the MLIR PR. Please drop this check.

Comment (krzysz00) [reply src/solver/mlir_common.cpp:88]: ... yeah, we do have a rather strict layout filter, and so could filter out 3D that way. Sure.

**Comment by jerryyin** [comment]
Considering that e4946eb has a full passing record whereas regression appears on f0457a2. It is likely a regression has already been introduced in MLIR end in the `miopen-dialect` branch. Could you prioritize a fix or revert on the MLIR side?

**Review by jerryyin — APPROVED** [review]
No further opinions provided that all MLIR CI continue to pass.

File: src/solver/mlir_common.cpp [hunk 4]
```diff
@@ -68,11 +86,11 @@ std::string ConstructBuildOptions(const ConvolutionContext& ctx,
         std::string(" --arch ") + ctx.GetStream().GetDeviceName() +
         std::string(" --groupsize ") + std::to_string(CI::GetGroupCountG(ctx)) +
         std::string(" --fil_layout ") + fil_layout +
-        std::string(" --fil_type ") + data_type +
+        std::string(" --fil_type ") + DTypeName(ctx.weights_data_type) +
```
Comment (atamazov) [inline src/solver/mlir_common.cpp:89]: [future] This is cumbersome. Use ostringstream, them return `.str()`.

**Review by atamazov — APPROVED** [review]
This can be merged. Please note that CI performs only basic testing, so I highly recommend supplementing CI testing with rigorous manual validation.

Please let me know if you would like me to merge this as is.

File: src/solver/mlir_common.cpp (hunk 4 above)
Comment (krzysz00) [reply src/solver/mlir_common.cpp]: Agreed and fixed

File: src/solver/mlir_common.cpp [hunk 5]
```diff
@@ -41,6 +43,23 @@ std::string InsertGToLayout(const std::string& layout, char dim)
...
+    {
+    case miopenHalf: return "fp16";
+    case miopenFloat: return "fp32";
+    case miopenDouble: return "fp64";
+    case miopenBFloat16: return "bf16";
+    case miopenInt32: return "i32";
+    case miopenInt8: return "i8";
+    case miopenInt8x4: return "i8x4";
+    }
+    assert(false); // All cases should be covered before here
```
[the remark this answers is not in the document]
Comment (krzysz00) [reply src/solver/mlir_common.cpp]: Thanks for the FYI - fixed

**Comment by krzysz00** [comment]
@atamazov If you have no other comments, please feel free to merge

File: src/solver/mlir_common.cpp [hunk 6]
```diff
@@ -41,56 +45,74 @@ std::string InsertGToLayout(const std::string& layout, char dim)
...
+    {
+    case miopenHalf: return "fp16";
+    case miopenFloat: return "fp32";
+    case miopenDouble: return "fp64";
+    case miopenBFloat16: return "bf16";
+    case miopenInt32: return "i32";
+    case miopenInt8: return "i8";
+    case miopenInt8x4: return "i8x4";
+    }
+    MIOPEN_THROW("Internal error: value outside of datatype enum");
```
Comment (atamazov) [inline src/solver/mlir_common.cpp:60]: [Recommendation] Here's what I meant:
```suggestion
    MIOPEN_THROW(miopenStatusInternalError, "Value outside of datatype enum");
```

File: src/solver/mlir_common.cpp [hunk 7]
```diff
@@ -24,10 +24,14 @@
  *
  *******************************************************************************/
 
+#include "miopen/miopen.h"
```
Comment (atamazov) [inline src/solver/mlir_common.cpp:27]:
```suggestion
#include <miopen/miopen.h>
```

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/solver/mlir_common.cpp (hunk 7 above)
Comment (atamazov) [reply src/solver/mlir_common.cpp:27]: This needs to be fixed, sorry.

**Review by atamazov — APPROVED** [review]
LGTM!

**Pull request merged**